### ISSUE TITLE
"edge" property broken on all builds

### ISSUE THREAD

**Issue by mboudreau** [travis-ci/dpl#687]
Most of our lambda deployments are using the "edge" property for the deployment, which points to my own fork of DPL with configuration for the latest lambda.  After trying a deployment this morning, it seems like the current deployment is ignoring the edge gem install altogether.  Here's an example of what it's doing:

```
Building dpl gem locally with source mboudreau/dpl and branch master
$ git clone https://github.com/mboudreau/dpl mboudreau/dpl
Cloning into 'mboudreau/dpl'...
remote: Counting objects: 8054, done.
remote: Compressing objects: 100% (6/6), done.
remote: Total 8054 (delta 4), reused 0 (delta 0), pack-reused 8048
Receiving objects: 100% (8054/8054), 2.19 MiB | 0 bytes/s, done.
Resolving deltas: 100% (4896/4896), done.
0.01s
0.02s$ git checkout master
Already on 'master'
Your branch is up-to-date with 'origin/master'.
0.01s$ git show-ref -s HEAD
542de9b0dc31ae9569b374b84dacb8eeb8b35e22
6.11s$ rvm $(travis_internal_ruby) --fuzzy do ruby -S gem build dpl.gemspec
Latest version currently installed. Aborting.
WARNING:  pessimistic dependency on rspec (~> 3.0.0, development) may be overly strict
  if rspec is semantically versioned, use:
    add_development_dependency 'rspec', '~> 3.0', '>= 3.0.0'
WARNING:  open-ended dependency on rspec-its (>= 0, development) is not recommended
  if rspec-its is semantically versioned, use:
    add_development_dependency 'rspec-its', '~> 0'
WARNING:  open-ended dependency on rake (>= 0, development) is not recommended
  if rake is semantically versioned, use:
    add_development_dependency 'rake', '~> 0'
WARNING:  open-ended dependency on coveralls (>= 0, development) is not recommended
  if coveralls is semantically versioned, use:
    add_development_dependency 'coveralls', '~> 0'
WARNING:  See http://guides.rubygems.org/specification-reference/ for help
  Successfully built RubyGem
  Name: dpl
  Version: 1.8.41.travis.220.1
  File: dpl-1.8.41.travis.220.1.gem
6.95sLatest version currently installed. Aborting.
Successfully installed dpl-1.8.41.travis.220.1
1 gem installed
Latest version currently installed. Aborting.
dpl.1
Installing deploy dependencies
Fetching: rubyzip-1.2.1.gem (100%)
Successfully installed rubyzip-1.2.1
1 gem installed
dpl.2
Preparing deploy
Using Access Key: ****************RCMA
dpl.3
Deploying application
/home/travis/.rvm/gems/ruby-2.2.7/gems/aws-sdk-core-2.10.38/lib/seahorse/client/plugins/raise_response_errors.rb:15:in `call': User: arn:aws:iam::047651431481:user/Travis_Orca is not authorized to perform: lambda:ListFunctions (Aws::Lambda::Errors::AccessDeniedException)
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/aws-sdk-core-2.10.38/lib/aws-sdk-core/plugins/jsonvalue_converter.rb:20:in `call'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/aws-sdk-core-2.10.38/lib/aws-sdk-core/plugins/idempotency_token.rb:18:in `call'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/aws-sdk-core-2.10.38/lib/aws-sdk-core/plugins/param_converter.rb:20:in `call'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/aws-sdk-core-2.10.38/lib/aws-sdk-core/plugins/response_paging.rb:26:in `call'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/aws-sdk-core-2.10.38/lib/seahorse/client/plugins/response_target.rb:21:in `call'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/aws-sdk-core-2.10.38/lib/seahorse/client/request.rb:70:in `send_request'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/aws-sdk-core-2.10.38/lib/seahorse/client/base.rb:207:in `block (2 levels) in define_operation_methods'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/dpl-1.8.42/lib/dpl/provider/lambda.rb:30:in `push_app'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/dpl-1.8.42/lib/dpl/provider.rb:154:in `block in deploy'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/dpl-1.8.42/lib/dpl/cli.rb:41:in `fold'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/dpl-1.8.42/lib/dpl/provider.rb:154:in `deploy'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/dpl-1.8.42/lib/dpl/cli.rb:32:in `run'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/dpl-1.8.42/lib/dpl/cli.rb:7:in `run'
	from /home/travis/.rvm/gems/ruby-2.2.7/gems/dpl-1.8.42/bin/dpl:5:in `<top (required)>'
	from /home/travis/.rvm/gems/ruby-2.2.7/bin/dpl:23:in `load'
	from /home/travis/.rvm/gems/ruby-2.2.7/bin/dpl:23:in `<main>'
failed to deploy
```

As you can see, it is in fact building the gem, but then seems to be ignoring the installation of it because the newer one is already available:

```
Latest version currently installed. Aborting.
Successfully installed dpl-1.8.41.travis.220.1
1 gem installed
```
The error is coming from `dpl-1.8.42` and not the newly installed one, which removes the need for ListFunctions.  This used to work, so I'm guessing either something changed in how travis handles locally built gems or that it's always worked this way, but a new version of DPL is superseding my fork.

I think the workaround is to just merge in the latest from the upstream hopefully, but it's still very confusing that it works this way, which probably shouldn't.

**Comment by BanzaiMan** [comment]
I think this is a problem with your fork. aws-sdk 3.0.0 was released recently, and it broke a few things (cf. #685). A fix was deployed as 1.8.42. Please rebase or merge `master` to reflect this.

**Comment by mboudreau** [comment]
I did, and it did fix it, however, I'm positive this is not an aws sdk
issue as you can clearly see in the logs that the gem used to deploy was
the newer one, and not my forked one.  If anything it might be a change on
how gem works behind the scenes when you try to install a local gem that's
deprecated by one already installed with a higher version number.  Not sure
what the correct fix for this is as I'm not much of a Ruby dev.

**Comment by BanzaiMan** [comment]
Ah, I see. So, your `dpl` gem was installed, but something else installed `1.8.42`, preventing your code to run.

Do you have a build log URL to share?

**Comment by mboudreau** [comment]
Not exactly, as far as I can read what gem is doing, 1.8.42 was installed
beforehand with deployment step (npm publish), then when it came to the
lambda deployment, I specify the edge branch I want it to use and then
builds the gem and tries to install it, but ignores it because 1.8.42 is
newer and already installed.

Sadly, I can't show you the build log as it's in a private repo.

**Comment by BanzaiMan** [comment]
@mboudreau How exactly did you fix it, then?

I don't think `npm publish` is a part of the normal workflow. If it is running, then, I suspect you might have another deployment provider (`npm`) defined which is installing `dpl` 1.8.42 before your fork runs.

**Comment by mboudreau** [comment]
@BanzaiMan Sorry, I'm not communicating clearly enough, let me try again.  My build currently has the following deployment steps:

- npm deployment, non edge
- s3 deployment, non edge
- Lambda deployment, edge using my git branch

When arriving at the lambda deployment, it tries to install an older version of dpl and ignores it because the newest version is available.  This is unexpected behavior as I'd expect "edge" would force install it no matter which version is on my branch.

I was able to work around this issue by fetching the latest from upstream repo (dpl) and merging it into my branch, which updated the version to the latest.

Is this clear?

**Comment by BanzaiMan** [comment]
Yup. It is indeed clear. And I think your analysis does agree with my second guess.

Your proposed solution might be a good way to get around this problem. I'll look into that soon. (In the mean time, of course, you can do that manually in `before_deploy`, though it would be a little time-consuming.)

**Comment by BanzaiMan** [comment]
Hmmm. To think, it is entirely possible that `edge` is actually *ahead* of the release, and the release stops working as a result. Seems that the only way to guarantee that `dpl` works is to uninstall it every time.

**Comment by mboudreau** [comment]
That's true. Can you do force install with gem?  But yeah, you'll have to
do it for every deploy item.
